### PR TITLE
Fixed #34803 -- Fixed queryset crash when filtering againts deeply nested OuterRef annotations.

### PR DESCRIPTION
**Pull request by Gnomeek**
fix issue [#34803](https://code.djangoproject.com/ticket/34803)

adding a new model to mock the minimal reproducible case in [thread discussion](https://code.djangoproject.com/ticket/34803)

### CODE REVIEW

**Review by shangxiao — APPROVED** [review]
Hi & welcome!  Nice.  Not sure about the employee -> manager -> secretary chain as it sorta breaks the child -> parent flow but it does the job.

We normally don't comment on what tests are doing but I think in this case it might not be obvious that the inner OuterRef refers to an outer pure OuterRef instead of a ResolvedOuterRef. A comment stating that this is different from other nested OuterRef tests may be helpful here? 🤔

File: tests/expressions/models.py [hunk 1]
```diff
@@ -6,8 +6,14 @@
...
 
 
+class Secretary(models.Model):
+    name = models.CharField(max_length=50)
+    based_in_eu = models.BooleanField(default=False)
+
+
 class Manager(models.Model):
     name = models.CharField(max_length=50)
+    secretary = models.ForeignKey(Secretary, models.CASCADE, null=True)
```
Comment (shangxiao) [inline tests/expressions/models.py:16]: I just realised we could do the same test by having this point to Employee, then we don't need the Secretary model.

Comment (Gnomeek) [reply tests/expressions/models.py:16]: well, Changing `Secretary` to `Employee` will lead to `used-before-assignment` error. Putting `Manager` class definition under `Employee` class definition will not fix it and vice versa...

Comment (shangxiao) [reply tests/expressions/models.py:16]: I don't know what a `used-before-assignment` error is but you can definitely point secretary to `Employee`. The problem then becomes choosing something to compare through the outer ref chain. You could simply use pks and then confirm the query doesn't result in an exception. Or we could just add the based in EU flag on Employee.

Comment (felixxm) [reply tests/expressions/models.py:16]: You can use 
```suggestion
    secretary = models.ForeignKey("Employee", models.CASCADE, null=True, related_name="secretaries")
```

Comment (Gnomeek) [reply tests/expressions/models.py:16]: lgtm, will give a shot

**Comment by felixxm** [comment]
@Gnomeek Thanks :+1: Welcome aboard :boat: 

I added a release note, and push small edits to tests.

**Pull request merged**